### ISSUE TITLE
Volume gets random Mat color ID

### ISSUE THREAD

**Issue by EgertKanep** [LuxCoreRender/LuxCore#426]
When using volume together with glass shader it does not get specified material id color. It should be red as is arch glass
![mat id color volume](https://user-images.githubusercontent.com/41080633/89581279-5fcc8680-d83f-11ea-8a27-3bdb3a563ced.JPG)

**Comment by Dade916** [comment]
A volume has its own Object ID like any other material. If there a scattering event (i.e. a reflection) inside the volume, the pixel should get the ID of the volume. Other effects than scattering, like absorption or emission, will not affect the Object ID AOV at all.

This is the way it is supposed to work, are you observing a different behavior ? If you post a simple test scene (i.e. a cube with a volume) showing the problem, I can look into it.

**Comment by EgertKanep** [comment]
Problem is specifically with material id color aov. Test file https://we.tl/t-jNGpYDRXJ3

I have specified material id to be 255 which should be pure red in material id color aov, but volume seems not to respect that setting

**Comment by Dade916** [comment]
The surface ID is:

`scene.materials.Material140708415325576.id = 255`

while the volume ID is:

`scene.volumes.140708410598920Volume.id = 3364224`

So the OBJECT ID AOV is a mix of both values because some time rays are reflected by the surface (i.e. ID=255) and some time are scattered by the volume (i.e. ID=3364224).

If you set the the material to NULL, it will be totally transparent so only the volume ID (ID=3364224) will be written on the AOV.

If you set the the material to GLASS (and the volume IOR to 1.0), only the material ID (ID=255) will be written on the AOV. I assume, this is what you were looking for.

@Theverat, I'm not sure if BlendLuxCore allows you to edit the volume ID (?).

**Comment by EgertKanep** [comment]
I am looking for a way to have my specified material id seen in material id color aov. If it is a mix of different Id's I have hard time masking.

**Comment by Dade916** [comment]
> I am looking for a way to have my specified material id seen in material id color aov. If it is a mix of different Id's I have hard time masking. I do not want to rerender with different settings to render just my masks

It is not possible if you use ARCHGLASS because the material is not "solid": some time it affects rays (and mark the OBJECT ID AOV), sometime it doesn't. Any other material will work.
ARCHGLASS has this uncommon behavior to work around some path tracing limitation.

**Comment by EgertKanep** [comment]
But currently arch glass gets correct Id as I have set up. Volume is the one that gets it random every time, is it not?

**Comment by Dade916** [comment]
Yes, if you use set both to the same value, you will get only one result (ID=255) however I don't know if you can set a volume ID in BlendLuxCore, I see there is a volume node editor but I don't know if you can set the volume ID there (?).

**Comment by EgertKanep** [comment]
There is no option for volume id there, so this has more to do with BlendLuxCore?

**Comment by Dade916** [comment]
Yes, @Theverat, BlendLuxCore should allow to set the volume ID, going to transfer this issue to BlendLuxCore.
